internal/server/config.go

```diff
@@ -523,57 +523,78 @@ func getKindFromUnstructured(data interface{}) string {
 	return ""
 }
 
-func (s Server) loadConfig(config Config) error {
+func (s *Server) loadConfig(db *gorm.DB, config Config) error {
 	if err := validator.New().Struct(config); err != nil {
 		return err
 	}
 
-	return s.db.Transaction(func(tx *gorm.DB) error {
-		// inject internal infra provider
-		config.Providers = append(config.Providers, Provider{
-			Name: models.InternalInfraProviderName,
-		})
+	return db.Transaction(func(tx *gorm.DB) error {
+		// load default infra provider
+		provider, err := s.loadProvider(tx, Provider{Name: models.InternalInfraProviderName})
+		if err != nil {
+			return err
+		}
 
-		config.Identities = append(config.Identities, Identity{
-			Name: models.InternalInfraConnectorIdentityName,
-		})
+		// load default connector identity
+		connector, err := s.loadIdentity(tx, Identity{Name: models.InternalInfraConnectorIdentityName}, provider)
+		if err != nil {
+			return err
+		}
 
-		config.Grants = append(config.Grants, Grant{
+		// load default connector grant
+		connectorGrant, err := s.loadGrant(tx, Grant{
 			User:     models.InternalInfraConnectorIdentityName,
 			Role:     models.InfraConnectorRole,
 			Resource: "infra",
 		})
-
-		if err := s.loadProviders(tx, config.Providers); err != nil {
-			return fmt.Errorf("load providers: %w", err)
+		if err != nil {
+			return err
 		}
 
-		// extract identities from grants and add them to identities
-		for _, g := range config.Grants {
-			switch {
-			case g.User != "":
-				config.Identities = append(config.Identities, Identity{Name: g.User})
-			case g.Machine != "":
-				logging.S.Warn("please update 'machine' grant to 'user', the 'machine' grant type is deprecated and will be removed in a future release")
-				config.Identities = append(config.Identities, Identity{Name: g.Machine})
+		// load and replace providers if specified in the config
+		if len(config.Providers) > 0 {
+			if err := s.loadProviders(tx, config.Providers, []uid.ID{provider.ID}); err != nil {
+				return err
 			}
 		}
 
-		if err := s.loadIdentities(tx, config.Identities); err != nil {
-			return fmt.Errorf("load identities: %w", err)
+		var ids []uid.ID
+
+		// load and replace grants if specified in the config
+		if len(config.Grants) > 0 {
+			for _, g := range config.Grants {
+				if g.Group != "" || g.User == "" {
+					continue
+				}
+
+				id, err := s.loadIdentity(tx, Identity{Name: g.User}, provider)
```

Comment:
The implication here is users/groups created through grants will never be removed, even if the grants are removed, since it's not calling `DeleteIdentities`